File: 0503-next-greater-element-ii/0503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];
        Arrays.fill(ans, -1);

        Deque<Integer> st = new ArrayDeque<>();

        for(int i = 0; i < 2*n; i++){
            while(!st.isEmpty() && nums[st.peek()] < nums[i%n])
            {
                ans[st.pop()] = nums[i%n];
            }

            if(i < n)
                st.push(i);
        }

        return ans;
    }
}

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);

        stack<int> st;
        int n = nums.size();
        for(int i = n-1; i >=0; i--){
            st.push(nums[i]);
        }
        vector<int> ans(n);
        for(int i = n-1; i >=0; i--){
            while(!st.empty() && st.top() <= nums[i]){
                st.pop();
            }

            if(st.empty()){
                ans[i] = -1;
            }
            else{
                ans[i] = st.top();
            }

            st.push(nums[i]);
        }

        return ans;
    }
};